### PR TITLE
Task_7

### PR DESCRIPTION
**Pull request by evTyurin**

### CODE REVIEW

File: src/main/java/com/senlainc/warsaw/tyurin/dao/CraftsmanDAO.java [hunk 1]
```diff
@@ -15,11 +17,15 @@ public class CraftsmanDAO implements ICraftsmanDAO{
     private List<Craftsman> craftsmen;
     private CsvReader csvReader;
     private CsvWriter csvWriter;
+    private JsonReader jsonReader;
```
Comment (mrseneca) [inline src/main/java/com/senlainc/warsaw/tyurin/dao/CraftsmanDAO.java]: I missed it last time I reviewed your task. my bad
Dao layer work with data, no need to bring here export/import work. 
Move this to service layer, work with export there and pass in dao already prepared entities

File: src/main/java/com/senlainc/warsaw/tyurin/service/CraftsmanService.java [hunk 2]
```diff
@@ -144,14 +145,35 @@ public void importCraftsmen() {
...
+    public void exportCraftsmenToCsv() {
 
         List<Craftsman> craftsmen = craftsmanDAO
                 .getCraftsmen()
                 .stream()
                 .sorted(Comparator.comparing(Craftsman::getId))
                 .collect(Collectors.toList());
 
-        craftsmanDAO.exportCraftsmen(craftsmen, Constants.PATH_TO_CRAFTSMEN_CSV);
+        craftsmanDAO.exportCraftsmenToCsv(craftsmen, Constants.PATH_TO_CRAFTSMEN_CSV);
```
Comment (mrseneca) [inline src/main/java/com/senlainc/warsaw/tyurin/service/CraftsmanService.java]: Same here, Entity should only contain info about itself and be able to provide api for crud-operations. 
And not it's also know how to write/read itself from/to external source.
Move this logic to service layer

**Review by mrseneca — CHANGES_REQUESTED** [review]
Let's change a bit code from previous task and some code from this task. Move export/import/json-serialization from dao to service layer. It will be better for future development

**Comment by evTyurin** [comment]
> Let's change a bit code from previous task and some code from this task. Move export/import/json-serialization from dao to service layer. It will be better for future development

Done

File: src/main/java/com/senlainc/warsaw/tyurin/util/Constants.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
+package com.senlainc.warsaw.tyurin.util;
+
+public class Constants {
+    public static String PATH_TO_CRAFTSMEN_CSV = "src\\main\\resources\\craftsmen.csv";
```
Comment (mrseneca) [inline src/main/java/com/senlainc/warsaw/tyurin/util/Constants.java:4]: When you use relative paths, you may experience problems running the application on other operating systems. Your property files are already in resources, which means they are in the classpath. You can specify paths taking into account the classpath, i.e. PATH_TO_PROPERTIES = "application.properties";

Additionally, you just need to change how you read property files, don't refer to them as files, but use the classloader to refer to them as resources

Comment (mrseneca) [inline src/main/java/com/senlainc/warsaw/tyurin/util/Constants.java:4]: When you use relative paths, you may experience problems running the application on other operating systems. Your property files are already in resources, which means they are in the classpath. You can specify paths taking into account the classpath, i.e. PATH_TO_PROPERTIES = "application.properties";

Additionally, you just need to change how you read property files, don't refer to them as files, but use the classloader to refer to them as resources

File: src/main/java/com/senlainc/warsaw/tyurin/util/propertyHandlers/PropertyReader.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+    public static PropertyReader getInstance() {
+        if (INSTANCE == null) {
+            INSTANCE = new PropertyReader();
+        }
+        return INSTANCE;
+    }
+
+    public String readProperties(String propertyName, String path) {
+        try (InputStream input = new FileInputStream(path)) {
```
Comment (mrseneca) [inline src/main/java/com/senlainc/warsaw/tyurin/util/propertyHandlers/PropertyReader.java]: `InputStream input = PropertyReader.class.getClassLoader().getResourceAsStream(path)`

You can do it like this (check it by yourself, i didn't tested it)

File: src/main/java/com/senlainc/warsaw/tyurin/util/propertyHandlers/PropertyReader.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    public static PropertyReader getInstance() {
+        if (INSTANCE == null) {
+            INSTANCE = new PropertyReader();
+        }
+        return INSTANCE;
+    }
+
+    public String readProperties(String propertyName, String path) {
```
Comment (mrseneca) [inline src/main/java/com/senlainc/warsaw/tyurin/util/propertyHandlers/PropertyReader.java]: Also there is some performance issue - every time you try to get some property you rereading property file, i think it unnecessary, make it read property one time, and then get them from Property class

**Review by mrseneca — CHANGES_REQUESTED** [review]
Almost done. Just few tunes

File: resources/craftsmen.json [hunk 6]
```diff
@@ -0,0 +1 @@
+[{"id":1,"name":"bob","surname":"smith"},{"id":3,"name":"stan","surname":"smith"},{"id":6,"name":"bob","surname":"smith"},{"id":2,"name":"john","surname":"pops"}]
```
Comment (mrseneca) [inline resources/craftsmen.json]: Just curious. Why you need "red" resource folder when you already have "green" one?
изображение

If there is no any idea behind it, please move all source files to a proper one folder

File: src/main/java/com/senlainc/warsaw/tyurin/util/propertyHandlers/PropertyReader.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+            INSTANCE = new PropertyReader();
+        }
+        return INSTANCE;
+    }
+
+    private void readProperties() {
+        try (InputStream input = PropertyReader.class.getClassLoader().getResourceAsStream(Constants.PATH_TO_PROPERTIES)) {
+            System.out.println("readProperties");
+            properties.load(input);
+            propertiesMap.putAll(properties.entrySet()
```
Comment (mrseneca) [inline src/main/java/com/senlainc/warsaw/tyurin/util/propertyHandlers/PropertyReader.java]: Why create additional Object for properties? You already have Properties class here. Just use properties.getProperty(key).
No need for additional object

**Review by mrseneca — CHANGES_REQUESTED** [review]
Only two small things

File: resources/craftsmen.json (hunk 6 above)
Comment (evTyurin) [reply resources/craftsmen.json]: As you said before use of relative paths may cause problems running the application on other operating systems. So the idea was to simplify relative path by using only one folder from root folder

Comment (mrseneca) [reply resources/craftsmen.json]: Yeah, that's right, but there 2 folders rn. You can store all files in one folder under main package (where application.properties stored rn)

**Review by mrseneca — APPROVED** [review]
Good job

**Pull request merged**